### PR TITLE
PWX-21175: handle kernel crash in fastpath + compilation fix

### PR DESCRIPTION
**Pull request by sulakshm**
**What this PR does / why we need it**:

In 2.10.0 branch

**Which issue(s) this PR fixes** (optional)
Closes #
https://portworx.atlassian.net/browse/PWX-21201 - fix compilation issue
https://portworx.atlassian.net/browse/PWX-21175 - fix kernel crash

**Special notes for your reviewer**:
This change passes the automation test consistently.
Fixes two issues.
1/ kernel crash
2/ compilation issue.

PASSED last 6 consecutive runs in the automation test.

### CODE REVIEW

File: pxd.c [hunk 1]
```diff
@@ -1024,13 +1024,14 @@ static void pxd_rq_fn(struct request_queue *q)
 		fproot = &req->fproot;
 		fp_root_context_init(fproot);
 
+#ifdef __PX_FASTPATH__
 		if (pxd_dev->fp.fastpath) {
 			// route through fastpath
-			INIT_WORK(&fproot->work, fp_handle_io);
```
Comment (sulakshm) [inline pxd.c:1029]: since default action on work queue is processing IO, init work is moved to fp_root_context_init() method.

File: pxd.c [hunk 2]
```diff
@@ -1024,13 +1024,14 @@ static void pxd_rq_fn(struct request_queue *q)
 		fproot = &req->fproot;
 		fp_root_context_init(fproot);
 
+#ifdef __PX_FASTPATH__
```
Comment (sulakshm) [inline pxd.c:1027]: introduce compile time options to remove fastpath support and add stubs as necessary.

File: pxd.c [hunk 3]
```diff
@@ -1497,16 +1500,21 @@ ssize_t pxd_remove(struct fuse_conn *fc, struct pxd_remove_out *remove)
...
 		mutex_lock(&pxd_dev->disk->queue->sysfs_lock);
 
 		QUEUE_FLAG_SET(QUEUE_FLAG_DYING, pxd_dev->disk->queue);
 
-        mutex_unlock(&pxd_dev->disk->queue->sysfs_lock);
+		mutex_unlock(&pxd_dev->disk->queue->sysfs_lock);
+#else
+		blk_mq_freeze_queue(pxd_dev->disk->queue);
+		blk_set_queue_dying(pxd_dev->disk->queue);
+#endif
```
Comment (sulakshm) [inline pxd.c:1512]: use exported API to control block device queue state.

File: pxd.c [hunk 4]
```diff
@@ -1497,16 +1500,21 @@ ssize_t pxd_remove(struct fuse_conn *fc, struct pxd_remove_out *remove)
...
 	}
 
 	spin_unlock(&pxd_dev->lock);
 
-	pxd_fastpath_cleanup(pxd_dev);
+	disableFastPath(pxd_dev, false);
```
Comment (sulakshm) [inline pxd.c:1517]: just disable fastpath when device remove ioctl, do IO work queue cleanup when the pxd device remove handler gets called as the final action.

File: pxd.c [hunk 5]
```diff
@@ -1639,6 +1647,30 @@ static int __pxd_update_path(struct pxd_device *pxd_dev, struct pxd_update_path_
 	return pxd_init_fastpath_target(pxd_dev, update_path);
 }
 
+static void _pxd_setup(struct pxd_device *pxd_dev, bool enable)
```
Comment (sulakshm) [inline pxd.c:1650]: moved these 2 functions into pxd.c from pxd_fastpath.c

File: pxd.c [hunk 6]
```diff
@@ -1927,6 +1959,7 @@ static ssize_t pxd_debug_store(struct device *dev,
 			struct device_attribute *attr,
 		        const char *buf, size_t count)
 {
+#ifdef __PX_FASTPATH__
```
Comment (sulakshm) [inline pxd.c:1962]: block all the debug actions if fastpath is not supported.

File: pxd_bio_blkmq.c [hunk 7]
```diff
@@ -1,6 +1,6 @@
 // enable this only if the px block device IO is
 // registered through blkmq
-#ifdef __PXD_BIO_BLKMQ__
+#if defined __PXD_BIO_BLKMQ__ && defined __PX_FASTPATH__
```
Comment (sulakshm) [inline pxd_bio_blkmq.c:3]: compile time enable code.

File: pxd_bio_blkmq.c [hunk 8]
```diff
@@ -256,7 +257,9 @@ static int prep_root_bio(struct fp_root_context *fproot) {
                 return 0;
         }
 
-        rq_for_each_segment(bv, rq, rq_iter) nr_bvec++;
+        if (!specialops)
+                rq_for_each_segment(bv, rq, rq_iter) nr_bvec++;
+
```
Comment (sulakshm) [inline pxd_bio_blkmq.c:262]: This is the root cause of the crash. For discard IOs, the above function is indeterminate. Ideally should return zero vectors, but starting from 5.7 kernels discards/write same can carry special payloads. For px virtual device, no additional payload is needed. Compute bvecs only if not a special operation currently only discard.

File: pxd_bio_makereq.c [hunk 9]
```diff
@@ -1,6 +1,6 @@
 // enable this only if the px block device IO is
 // registered through make_request() fn.
-#ifdef __PXD_BIO_MAKEREQ__
+#if defined __PXD_BIO_MAKEREQ__ && defined __PX_FASTPATH__
```
Comment (sulakshm) [inline pxd_bio_makereq.c:3]: block code on compilation action

File: pxd_fastpath.c [hunk 10]
```diff
@@ -34,30 +35,6 @@ void pxd_abortfailQ(struct pxd_device *pxd_dev)
 	spin_unlock_irqrestore(&pxd_dev->fp.fail_lock, flags);
 }
 
-static void _pxd_setup(struct pxd_device *pxd_dev, bool enable)
```
Comment (sulakshm) [inline pxd_fastpath.c:37]: moved to pxd.c. These function are always needed.

File: pxd_fastpath.h [hunk 11]
```diff
@@ -65,6 +65,10 @@ struct pxd_fastpath_extension {
 	atomic_t nerror; // [global] total IO error
 };
 
+#ifndef __PX_FASTPATH__
+#include "pxd_fastpath_stub.h"
```
Comment (sulakshm) [inline pxd_fastpath.h:69]: introduce stub functions if fastpath is not enabled.

**Review by prabirpaul — APPROVED** [review]

**Pull request merged**